### PR TITLE
Update CsvMonumentConverter and InitialDataLoadApplication for New Fields

### PR DESCRIPTION
**Pull request by njd9419**
Finally getting this into PR. There were a few things that came up that prevented me from getting it up sooner.

This PR:
- Parses the additional columns in the Initial Dataset Spreadsheet to fill in the new attributes we discussed: References, Contributions, Inscriptions and Tags
- Adds the groundwork for Monument (and the broader Model) validation and exposes it in a manual manner if needed
- Adds a bunch of unit and integration tests around the new things implemented

When you go to run this on to load your local database, be sure to download and use the Initial Dataset.csv file that is in the Drive, inside the Data Sets from Dr. Decker folder. This CSV file is scrubbed into a almost 100% acceptable state so that the data we have is valid.

### CODE REVIEW

File: src/main/java/com/monumental/models/Contribution.java [hunk 1]
```diff
@@ -18,6 +19,7 @@
 public class Contribution extends Model implements Serializable {
 
     @Column(name = "submitted_by")
+    @NotNull(groups = NewOrExisting.class, message = "Submitted By can not be null")
```
Comment (njd9419) [inline src/main/java/com/monumental/models/Contribution.java]: `@NotNull` is one of the JPA constraint annotations and allows for a custom message when violated.

File: src/main/java/com/monumental/models/Model.java [hunk 2]
```diff
@@ -4,14 +4,38 @@
...
+    // Use this group to signify validation that occurs on record creation or after record creation
+    public interface NewOrExisting extends New, Existing {
+
+    }
+
+    // Use this group to signify validation of the Id field of the record
+    public interface Identity {
+
+    }
+
```
Comment (njd9419) [inline src/main/java/com/monumental/models/Model.java]: These are groups we can use on our `@NotNull` and other JPA constraint annotations to signify when we want the validation to occur.

File: src/main/java/com/monumental/models/Monument.java [hunk 3]
```diff
@@ -258,4 +267,82 @@ private String generateDescription() {
...
+        this.references.add(reference);
+    }
+
+    /**
+     * Encapsulates the logic to validate a Monument object
+     * Use this method to manually run validation in lieu of a @Valid Spring annotation
+     * @return List<String> - List of ConstraintViolation messages, if any
+     * @return null - If there are no ConstraintViolations
+     */
+    public List<String> validate() {
```
Comment (njd9419) [inline src/main/java/com/monumental/models/Monument.java]: This is a method that we can use to manually validate a Monument object instead of relying on an `@Valid` annotation to do it for us.

Since Java doesn't have `out`, the return of this method isn't what I'd want it to be (returns a boolean and outs a List of exception messages) so it's on the caller to get if what is returned is null or not to determine if the Monument is valid.

File: src/main/java/com/monumental/services/ModelService.java [hunk 4]
```diff
@@ -281,6 +281,39 @@ public void delete(List<Integer> ids) throws HibernateException {
         return records;
     }
```
Comment (njd9419) [inline src/main/java/com/monumental/services/ModelService.java]: This is a generic method we can use in the future to make queries with `WHERE` clauses on our tables without having to repeat all of the other junk.

File: src/main/java/util/InitialDataLoadApplication.java [hunk 5]
```diff
@@ -35,23 +41,61 @@ public static void main(String[] args) {
...
+                if (r.getTags() != null) {
+                    for (Tag t : r.getTags()) {
+                        try {
+                            tagService.insert(t);
+                            tagInsertedCount++;
+                        } catch (ConstraintViolationException e) {
+                            // If a ConstraintViolationException is caught, there's a duplicate Tag in the list we are trying
+                            // to insert
+                            // Instead, get the ID of the already existing Tag and update it
+                            // Assume there is only 1 Tag with that Name due to the Unique Constraint
```
Comment (njd9419) [inline src/main/java/util/InitialDataLoadApplication.java]: Having to catch a `ConstraintViolationException`, grabbing the duplicate's ID, setting it then updating isn't ideal. Perhaps in the future we can change `insert()` to encapsulate this logic so the caller doesn't have to worry about it.

File: src/main/java/com/monumental/models/Monument.java (hunk 3 above)
Comment (benvogler) [reply src/main/java/com/monumental/models/Monument.java]: Instead of using `out` you could just make a class (within this class) as a data holder and return that

Comment (njd9419) [reply src/main/java/com/monumental/models/Monument.java]: Now makes use of `Monument.ValidationResult`

File: src/main/java/com/monumental/models/Model.java [hunk 6]
```diff
@@ -4,14 +4,38 @@
...
+    // Use this group to signify validation that occurs on record creation or after record creation
+    public interface NewOrExisting extends New, Existing {
+
+    }
+
+    // Use this group to signify validation of the Id field of the record
+    public interface Identity {
```
Comment (benvogler) [inline src/main/java/com/monumental/models/Model.java]: I know I probably wrote this, but I really don't understand what this is for

File: src/main/java/com/monumental/models/Monument.java [hunk 7]
```diff
@@ -52,23 +60,23 @@
     @Column(name = "description")
     private String description;
 
-    @Column(name = "inscription")
+    @Column(name = "inscription", length = 1024)
```
Comment (benvogler) [inline src/main/java/com/monumental/models/Monument.java]: what does this do? make the max length 1024 chars? if so that's probably too short

File: src/main/java/com/monumental/models/Monument.java [hunk 8]
```diff
@@ -249,7 +257,8 @@ private String generateDescription() {
                 Reference firstReference = this.references.get(0);
 
                 if (firstReference != null && firstReference.getUrl() != null) {
-                    description += " You may find further information about this monument at: " + firstReference.getUrl();
+                    description += " You may find further information about this monument or memorial at: " +
+                            firstReference.getUrl();
```
Comment (benvogler) [inline src/main/java/com/monumental/models/Monument.java]: Sort of related to the image gallery we may want a way to signify a primary reference rather than just take the oldest

File: src/main/java/com/monumental/models/Monument.java [hunk 9]
```diff
@@ -258,4 +267,108 @@ private String generateDescription() {
...
+
+    /**
+     * Adds a Contribution to the List
+     * Will make a new ArrayList if this.contributions is null
+     * Checks if a Contribution has already been added to this.contributions with the same date and name
+     * Will do nothing if the specified Contribution has a null date or null/empty submittedBy
+     * and does nothing if so
+     * @param contribution - Contribution to add to the List
+     */
+    public void addContribution(Contribution contribution) {
```
Comment (benvogler) [inline src/main/java/com/monumental/models/Monument.java]: These are covered by the `NotNull` and such not aren't they? do we still need these? I think this is bad design if we do need them

File: src/main/java/com/monumental/models/Monument.java [hunk 10]
```diff
@@ -258,4 +267,108 @@ private String generateDescription() {
...
+
+        this.references.add(reference);
+    }
+
+    /**
+     * Encapsulates the logic to validate a Monument object
+     * Use this method to manually run validation in lieu of a @Valid Spring annotation
+     * @return ValidationResult - ValidationResult object representing the result of the validation
+     */
+    public ValidationResult validate() {
```
Comment (benvogler) [inline src/main/java/com/monumental/models/Monument.java]: Why is this needed?

File: src/main/java/com/monumental/services/ModelService.java [hunk 11]
```diff
@@ -281,6 +281,39 @@ public void delete(List<Integer> ids) throws HibernateException {
...
 
+    /**
+     * Generic method to build a "FROM ... WHERE ..." query
+     * Assumes the FROM clause is the name of the T generic class
+     * @param whereParameterName - Name of the column to use in the WHERE clause
+     * @param whereParameter - Used to set the parameter in the WHERE clause
+     * @return List<T> - the List of T returned by the query
+     */
+    @SuppressWarnings("unchecked")
+    public List<T> getFromWhere(String whereParameterName, Object whereParameter) {
```
Comment (benvogler) [inline src/main/java/com/monumental/services/ModelService.java]: This should also have an option to initialize all the lazy loaded lists like in the other gets

File: src/main/java/com/monumental/services/TagService.java [hunk 12]
```diff
@@ -8,11 +8,36 @@
 @Service
 public class TagService extends ModelService<Tag> {
 
+    /**
+     * Public constructor for TagService
+     * Use when NOT injecting SessionFactoryService via Spring
+     * @param sessionFactoryService - instance of SessionFactoryService to use for initialization
+     */
+    public TagService(SessionFactoryService sessionFactoryService) {
```
Comment (benvogler) [inline src/main/java/com/monumental/services/TagService.java]: If all of the services are going to have this then it should just be done in the `ModelService` itself

File: src/main/java/com/monumental/services/ModelService.java (hunk 11 above)
Comment (benvogler) [reply src/main/java/com/monumental/services/ModelService.java]: Check out hibernate `Criteria` objects https://docs.jboss.org/hibernate/orm/3.5/javadocs/org/hibernate/Criteria.html

Instead of having this function be limited to one equals, you could instead have `ModelService.createCriteria`, then you do all your `.add`s to it and pass the completed `Criteria` object to a `ModelService.executeQuery(Criteria criteria)`

File: src/main/java/util/InitialDataLoadApplication.java (hunk 5 above)
Comment (benvogler) [reply src/main/java/util/InitialDataLoadApplication.java]: You could override `doInsert` to achieve this pretty easily for one service

File: src/main/java/com/monumental/models/Model.java (hunk 6 above)
Comment (njd9419) [reply src/main/java/com/monumental/models/Model.java]: I added the `Identity` one. I was running into the issue that if I called the `validate` method and the `id` field was in the `NewOrExisting` group I would get a `ConstraintViolationException`.

File: src/main/java/util/InitialDataLoadApplication.java (hunk 5 above)
Comment (njd9419) [reply src/main/java/util/InitialDataLoadApplication.java]: Don't feel like doing that in this PR, but we should in the future

File: src/main/java/com/monumental/models/Monument.java (hunk 7 above)
Comment (njd9419) [reply src/main/java/com/monumental/models/Monument.java]: That's just what I found to work for the initial data load.

I'll increase it to 2048.

File: src/main/java/com/monumental/models/Monument.java (hunk 10 above)
Comment (njd9419) [reply src/main/java/com/monumental/models/Monument.java]: It's used in the `InitialDataLoadApplication`, but they're probably not needed right? The validation should run on insert/update?

File: src/main/java/com/monumental/models/Monument.java (hunk 9 above)
Comment (njd9419) [reply src/main/java/com/monumental/models/Monument.java]: Well, asserting that they aren't null on insert/update of a Monument would be covered by `@NotNull`, but I think we still need a way to associate the Contribution with the Monument in-memory before the insert/update like we do for Tags, no?

File: src/main/java/com/monumental/models/Model.java (hunk 6 above)
Comment (benvogler) [reply src/main/java/com/monumental/models/Model.java]: Shouldn't `Existing` work? That's how we had it in 343

File: src/main/java/com/monumental/models/Monument.java (hunk 9 above)
Comment (benvogler) [reply src/main/java/com/monumental/models/Monument.java]: It seems like the entire point of these `addBlank` functions is to do a null check for you. Why not just initialize those lists in the `Monument` constructor if they're `null`?

```java
public Monument(...) {
    if (this.getContributions() == null) this.setContributions(new ArrayList<>());
    // etc.
}
```

File: src/main/java/com/monumental/models/Monument.java (hunk 10 above)
Comment (benvogler) [reply src/main/java/com/monumental/models/Monument.java]: I'm not sure. They run on the controller methods so maybe not. But maybe there is a way to use `@Valid` outside of controller method params.

File: src/main/java/com/monumental/services/ModelService.java [hunk 13]
```diff
@@ -282,33 +301,41 @@ public void delete(List<Integer> ids) throws HibernateException {
...
-     * @return List<T> - the List of T returned by the query
+     * Perform a get query with the specified criteria
+     * @param criteria - List of Criterion objects to apply to the query
+     * @param initializeLazyLoadedCollections - If true, loads all of the collections associated with T that are
+     *                                        normally lazy loaded
+     * @return List<T> - List of T returned by the query
      */
     @SuppressWarnings("unchecked")
-    public List<T> getFromWhere(String whereParameterName, Object whereParameter) {
+    public List<T> getWithCriteria(List<Criterion> criteria, boolean initializeLazyLoadedCollections) {
```
Comment (benvogler) [inline src/main/java/com/monumental/services/ModelService.java]: You should also add a `getWithCriteria(List<Criterion criteria> criteria)` that calls this method with `initializeLazyLoadedCollections` to false so that it's essentially an optional param. See the other get methods for reference

File: src/main/java/com/monumental/services/ModelService.java [hunk 14]
```diff
@@ -282,33 +301,41 @@ public void delete(List<Integer> ids) throws HibernateException {
...
         try {
             transaction = session.beginTransaction();
-            records = session.createQuery("FROM " + tableName + " WHERE " + whereParameterName + " = :whereParameter")
-                    .setParameter("whereParameter", whereParameter)
-                    .list();
+            Criteria c = session.createCriteria(this.getModelClass());
+
+            for (Criterion criterion : criteria) {
+                c.add(criterion);
+            }
```
Comment (benvogler) [inline src/main/java/com/monumental/services/ModelService.java]: beauty

File: src/main/java/com/monumental/services/MonumentService.java [hunk 15]
```diff
@@ -14,20 +14,8 @@
...
     public MonumentService(SessionFactoryService sessionFactoryService) {
-        this.sessionFactoryService = sessionFactoryService;
-    }
-
-    /**
-     * Public default constructor for MonumentService
-     */
-    public MonumentService() {
-
+        super(sessionFactoryService);
```
Comment (benvogler) [inline src/main/java/com/monumental/services/MonumentService.java]: Why do we even override the constructor here if it just calls super and returns?

File: src/main/java/com/monumental/services/TagService.java [hunk 16]
```diff
@@ -35,9 +26,13 @@ public TagService() {
...
+     * @param initializeLazyLoadedCollections - If true, initializes all of the Monuments associated with the Tags
      * @return List<Tag> - List of Tags with the specified name
      */
-    public List<Tag> getTagsByName(String name) {
-        return this.getFromWhere("name", name);
+    public List<Tag> getTagsByName(String name, boolean initializeLazyLoadedCollections) {
+        ArrayList<Criterion> criteria = new ArrayList<>();
+        criteria.add(Restrictions.eq("name", name));
+
+        return this.getWithCriteria(criteria, initializeLazyLoadedCollections);
```
Comment (benvogler) [inline src/main/java/com/monumental/services/TagService.java]: This is really smoove. There might be some other `get`s out there we can replace with this but I don't remember

File: src/main/java/com/monumental/services/ModelService.java (hunk 13 above)
Comment (benvogler) [reply src/main/java/com/monumental/services/ModelService.java]: Also this is super nitpicky but I kind of prefer that the method just be called `get` and accept the criteria param. I don't know why but I like them all having the same name

File: src/main/java/com/monumental/models/Monument.java (hunk 9 above)
Comment (njd9419) [reply src/main/java/com/monumental/models/Monument.java]: Fair enough you have provided all valid points. I'll fix that now.

File: src/main/java/com/monumental/services/ModelService.java (hunk 14 above)
Comment (njd9419) [reply src/main/java/com/monumental/services/ModelService.java]: I had to Google a couple of times which word was the plural and which was the singular.

File: src/main/java/com/monumental/services/MonumentService.java (hunk 15 above)
Comment (njd9419) [reply src/main/java/com/monumental/services/MonumentService.java]: Model is an abstract class so you can't do `MonumentService monumentService = new ModelService(sessionFactoryService)`. How else would you do this?

File: src/main/java/com/monumental/models/Model.java (hunk 6 above)
Comment (njd9419) [reply src/main/java/com/monumental/models/Model.java]: We talked on Slack:

Essentially the Id field should use `Existing` and `NewOrExisting` on most of the other columns. Then when validating (either in a Controller method parameter or manually), use `New` or `Existing`.

The only difference between `Existing` and `New` will be the ID.

Fixing this in the coming commit.

File: src/main/java/com/monumental/controllers/MonumentController.java [hunk 17]
```diff
@@ -20,6 +22,11 @@ public Monument createMonument(@RequestBody Monument monument) {
         return monument;
     }
 
+    @PostMapping("/api/monument/testValidation")
```
Comment (njd9419) [inline src/main/java/com/monumental/controllers/MonumentController.java]: This is just for testing and should be removed before merging.

**Comment by njd9419** [comment]
As far as I know this is good to go.

@benvogler are the changes you requested done?
